### PR TITLE
ENH: Adding DE2 instrument suite (partial) and constellation object

### PR DESCRIPTION
**Pull request by jklenzing**
# Description

The Eighties are back!

This pull adds support for the following instruments, all on board the Dynamics Explorer 2 (DE2) satellite:
- `de2_lang`: Langmuir Probe
- `de2_nacs`: Neutral Atmosphere Composition Spectrometer
- `de2_rpa`: Retarding Potential Analyzer
- `de2_wats`: Wind and Temperature Spectrometer

This also adds a constellation object for DE2.  Note this is not the complete list of available instruments, others are in various stages of implementation and will be part of future pulls. 

## Type of change

Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# How Has This Been Tested?

All instruments loaded into unit tests.  Additional checks in iPython:
```
import pysat
date = pysat.datetime(1983, 1, 1)

lang = pysat.Instrument(platform='de2', name='lang')
lang.download(date, date)
lang.load(date=date)

nacs = pysat.Instrument(platform='de2', name='nacs')
nacs.download(date, date)
nacs.load(date=date)

rpa = pysat.Instrument(platform='de2', name='rpa')
rpa.download(date, date)
rpa.load(date=date)

wats = pysat.Instrument(platform='de2', name='wats')
wats.download(date, date)
wats.load(date=date)
```
All instruments load both data and metadata as expected.

**Test Configuration**:
* OS 10.14.6
* Python 3.7.3

# Checklist:

- [x] Make sure you are merging into the ``develop`` (not ``master``) branch
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Add a note to ``CHANGELOG.md``, summarizing the changes

### CODE REVIEW

**Comment by jklenzing** [comment]
The main thing this branch needs is a way of checking the constellation object.  I don't quite know how to do this.

Once this is finalized, I'll create the twin pull into `develop-3`.  Waiting so as not to create too many tests in the queue.

**Comment by rstoneback** [comment]
Haven’t checked things out yet but nice. Thanks Jeff!

We will have to sort out what to do with constellation testing. The constellation object itself is only barebones and needs to be fleshed out. Could push some of this out 3.0. Not the pull itself.

**Comment by jklenzing** [comment]
Yeah, cleaning up my stale branches to get as much of the code out for review before things change too much.

I think pushing out the unit tests for the constellations to 3.0 is a good idea.  I just don't know what to do with the constellation now that I've set it up.

**Comment by rstoneback** [comment]
I have an ICON constellation that is also just currently hanging out. I’ve manually verified that it exists, and works, but I think automatic testing is limited to none.

I say leave it in, and well sort it out when we know better what to do with it.

**Comment by jklenzing** [comment]
In that case, this is ready for review.

File: pysat/instruments/de2_lang.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+"""
+
+from __future__ import print_function
+from __future__ import absolute_import
+
+import functools
+
+import pysat
+from .methods import nasa_cdaweb as cdw
```
Comment (rstoneback) [inline pysat/instruments/de2_lang.py]: from pysat.methods is more robust.

File: pysat/instruments/de2_nacs.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+"""
+
+from __future__ import print_function
+from __future__ import absolute_import
+
+import functools
+
+import pysat
+from .methods import nasa_cdaweb as cdw
```
Comment (rstoneback) [inline pysat/instruments/de2_nacs.py]: Same comment, for others as well.

**Comment by rstoneback** [comment]
Found a bug in nasa_cdaweb.py. Not strictly related to code changes you made, but I'm starting the documentation here since I found it checking this pull out.

rpa.remote_file_list()

UnboundLocalErrorTraceback (most recent call last)
<ipython-input-11-5cc467409375> in <module>()
----> 1 rpa.remote_file_list()

/Users/russellstoneback/git/pysat/pysat/_instrument.pyc in remote_file_list(self, year, month, day)
   1454 
   1455         return self._list_remote_rtn(self.tag, self.sat_id,
-> 1456                                      year=year, month=month, day=day)
   1457 
   1458     def remote_date_range(self, year=None, month=None, day=None):
/Users/russellstoneback/git/pysat/pysat/instruments/methods/nasa_cdaweb.pyc in list_remote_files(tag, sat_id, remote_site, supported_tags, user, password, fake_daily_files_from_monthly, two_digit_year_break, delimiter, year, month, day)
    527                 mask = mask & (stored_list.index.day == day)
    528 
--> 529     return stored_list[mask]

UnboundLocalError: local variable 'mask' referenced before assignment 

I expected a list of all available files.


Trying a single year didn't help,
rpa.remote_file_list(year=1983)

AttributeErrorTraceback (most recent call last)
<ipython-input-13-81ceeca92942> in <module>()
----> 1 rpa.remote_file_list(year=1983)

/Users/russellstoneback/git/pysat/pysat/_instrument.pyc in remote_file_list(self, year, month, day)
   1454 
   1455         return self._list_remote_rtn(self.tag, self.sat_id,
-> 1456                                      year=year, month=month, day=day)
   1457 
   1458     def remote_date_range(self, year=None, month=None, day=None):
/Users/russellstoneback/git/pysat/pysat/instruments/methods/nasa_cdaweb.pyc in list_remote_files(tag, sat_id, remote_site, supported_tags, user, password, fake_daily_files_from_monthly, two_digit_year_break, delimiter, year, month, day)
    521     # Downselect to user-specified dates, if needed
    522     if year is not None:
--> 523         mask = (stored_list.index.year == year)
    524         if month is not None:
    525             mask = mask & (stored_list.index.month == month)
AttributeError: 'Index' object has no attribute 'year'

**Comment by rstoneback** [comment]
Created issue #348

**Comment by rstoneback** [comment]
Code works as otherwise advertised. Thanks Jeff!

File: pysat/instruments/de2_lang.py (hunk 1 above)
Comment (jklenzing) [reply pysat/instruments/de2_lang.py]: This is persistent throughout the cdaweb instruments.  Perhaps a separate pull?

**Comment by rstoneback** [comment]
sure.

**Pull request merged**